### PR TITLE
Update dynamis keyboard

### PR DESCRIPTION
**Pull request by bbrfkr**
## Description

I updated the code of dynamis keyboard. In this work, the changes are the following:

- support remapping encoder
- update default keymap
- update readme

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by bbrfkr** [comment]
@drashna Could you check my PR?

**Comment by drashna** [comment]
> @drashna Could you check my PR?

This PR is 4 hours old, and has not been reviewed at all, yet.  In general, there is no need to tag me. (unless it's been a week or three).

In this context, this sort of pin is actually quite rude.  

Pull Request reviews are something that is done in our free time. We are not paid nor compensated for the time we spend reviewing, as it is a labor of love. As such, this means that it can take time for us to get to your Pull Request. Things like family, or life can get in the way of us getting to PRs, and burnout is a serious consideration. The QMK firmware repository averages 200 PRs opened and 200 PRs merged every month, so please have patience.


And to be clear, I'm not angry here, but this is behavior that we want to nip in the bud before it becomes a problem.

File: keyboards/bbrfkr/dynamis/dynamis.c [hunk 1]
```diff
@@ -17,17 +17,14 @@
 #include "dynamis.h"
 
 #ifdef ENCODER_ENABLE
-bool encoder_update_kb(uint8_t index, bool clockwise) {
-    if (!encoder_update_user(index, clockwise)) {
-        return false; 
+    bool encoder_update_user(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.c:20]: Please revert this change. 

This is absolutely not the way to handle this, and will cause issues.  Eg, if VIA is not enabled, this will simply not function.

Additionally, Encoder support should NOT be added to the keymap/layout macro, especially not at the keyboard level.  (VIA implementation should not bleed into keyboard code, but needs to be kept in the via keymap, when possible)

File: keyboards/bbrfkr/dynamis/dynamis.h [hunk 2]
```diff
@@ -23,7 +23,8 @@
       C10, C11, C12, C13, C14, C15, C16, C66, C65, C64, C63, C62, C61, C60, \
       C20, C21, C22, C23, C24, C25, C26, C76, C75, C74, C73, C72,      C70, \
       C30, C31, C32, C33, C34, C35, C36, C86, C85, C84, C83, C82, C81, C80,      C91, \
-      C40, C41, C42,      C43,      C44,           C45, C46, C96, C95,      C94, C93, C92 \
+      C40, C41, C42,      C43,      C44,           C45, C46, C96, C95,      C94, C93, C92, \
+      C100, C101 \
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.h:27]: These are sole for encoder support? 

If so, it needs to be removed.  VIA implementation should not be leaked into keyboard level code.

File: keyboards/bbrfkr/dynamis/readme.md [hunk 3]
```diff
@@ -1,13 +1,13 @@
 # dynamis
 
 ## general information
-![dynamis](https://i.imgur.com/q2C3D2ul.jpg)
+![dynamis](https://i.imgur.com/GRPE5hz.jpg)
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/readme.md:4]:
```suggestion
![dynamis](https://i.imgur.com/GRPE5hzh.jpg)
```

File: keyboards/bbrfkr/dynamis/dynamis.c (hunk 1 above)
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.c:20]: This won't actually work correctly, as tap_code (and tap_code16) only support basic keycodes (and modded basic keycodes). 

Ideally, this should be using the Encoder Map feature, but that's not supported by VIA yet. Though, the dynamic keymap feature (which is what powers VIA, really) actively supports it. 

https://docs.qmk.fm/#/feature_encoders?id=encoder-map

In the meanwhile, something like this should be used:
https://github.com/qmk/qmk_firmware/blob/b66b911fa20debfb6e49756dee57d22228f5f122/keyboards/dailycraft/stickey4/keymaps/via/keymap.c#L45-L115

This allows for proper handling of all keycodes, without special cases.

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by bbrfkr** [comment]
Dear drashna,

I apologize for my rude behavior. I mistaked the usage of @ tag...
I swear it will never happen again.

Thank you for pointing it out.

File: keyboards/bbrfkr/dynamis/dynamis.h (hunk 2 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.h:27]: Ok. So, How the behaviors of encoder for each layers should be implemented?

File: keyboards/bbrfkr/dynamis/dynamis.c (hunk 1 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.c]: Thank you. I moved the behavior of encoder for via to keymap level code.

Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.c]: Ok. I reverted the code and modified that this code run only when via is enabled.

**Comment by bbrfkr** [comment]
> These appear to be relatively significant changes to the matrix; does this represent new hardware that should in fact be managed in a rev2 folder?

No. This code is newly version firmware of dynamis keyboard, and support the same hardware.

File: keyboards/bbrfkr/dynamis/dynamis.h (hunk 2 above)
Comment (drashna) [reply keyboards/bbrfkr/dynamis/dynamis.h]: Either per the normal callback (encoder_update_kb) in the keyboard's C file.  This provides a default/fallback behavior if there is not a user function, and allows the user function to override the behavior if it is present. 

Alternatively, you can enable the encoder map feature at the keymap level.

**Comment by drashna** [comment]
> Dear drashna,
> 
> I apologize for my rude behavior. I mistaked the usage of @ tag... I swear it will never happen again.
> 
> Thank you for pointing it out.

No worries.  If it's been ... a while (eg, weeks), that's fine.  Also, if you exclude the "@", it won't tag that user. 

Eg, tagging a user is for getting their attention, but in this case, there is no need to get my attention (it's new, so I'll definitely see it when I go through PRs)

File: keyboards/bbrfkr/dynamis/dynamis.c [hunk 4]
```diff
@@ -16,18 +16,18 @@
 
 #include "dynamis.h"
 
-#ifdef ENCODER_ENABLE
+#ifndef VIA_ENABLE
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.c:19]: This should still be reverted. If you use the `encoder_update_user` function in the via keymap, and return false, it will prevent this code from being executed (that's what that first "if" is doing).  

Also, you wan to wrap the function in the encoder ifdef, so that if someone disables the feature, it doesn't error out during compiling. 
```suggestion
#ifdef ENCODER_ENABLE
```

File: keyboards/bbrfkr/dynamis/keymaps/via/keymap.c [hunk 5]
```diff
@@ -20,40 +20,79 @@ enum layer_names {
...
+            encoder_cw.pressed = true;
+            encoder_cw.time = (timer_read() | 1);
+            action_exec(encoder_cw);
+        } else {
+            encoder_ccw.pressed = true;
+            encoder_ccw.time = (timer_read() | 1);
+            action_exec(encoder_ccw);
+        }
+    }
+    return true;
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/keymaps/via/keymap.c:97]: Returning false will prevent the kb level code from executing.
```suggestion
    return false;
```

File: keyboards/bbrfkr/dynamis/dynamis.h (hunk 2 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.h]: I see. I updated to control with whether `layer_state == 0` or not.

File: keyboards/bbrfkr/dynamis/dynamis.c (hunk 4 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.c]: Thanks. I fixed it.

File: keyboards/bbrfkr/dynamis/keymaps/via/keymap.c (hunk 5 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/keymaps/via/keymap.c]: Oh, Ok. I fixed it too.

File: keyboards/bbrfkr/dynamis/dynamis.c [hunk 6]
```diff
@@ -16,18 +16,9 @@
 
 #include "dynamis.h"
 
-#ifdef ENCODER_ENABLE
+#ifndef ENCODER_ENABLE
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.c:19]: I suspect that you meant ifdef here, (and may be my fault)

```suggestion
#ifdef ENCODER_ENABLE
```

File: keyboards/bbrfkr/dynamis/dynamis.c [hunk 7]
```diff
@@ -16,18 +16,9 @@
...
-   if (index == 0) { /* First encoder */
-      if (clockwise) {
-         tap_code(KC_WH_D);
-      } else {
-         tap_code(KC_WH_U);
-      }
-   }
-   return false;
+    if (!encoder_update_user(index, clockwise)) { return false; }
+    return true;
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.c:22]: This this has a default behavior:
```suggestion
      if (clockwise) {
         tap_code(KC_WH_D);
      } else {
         tap_code(KC_WH_U);
      }
   }
   return true;
```

File: keyboards/bbrfkr/dynamis/dynamis.h [hunk 8]
```diff
@@ -35,7 +35,8 @@
         { C60, C61, C62, C63, C64, C65, C66 }, \
         { C70, KC_NO, C72, C73, C74, C75, C76 }, \
         { C80, C81, C82, C83, C84, C85, C86 }, \
-        { C90, C91, C92, C93, C94, C95, C96 } \
+        { C90, C91, C92, C93, C94, C95, C96 }, \
+        { KC_WH_D, KC_WH_U, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO } \
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.h:39]:
```suggestion
        { C90, C91, C92, C93, C94, C95, C96 } \
```

File: keyboards/bbrfkr/dynamis/dynamis.h [hunk 9]
```diff
@@ -55,5 +56,6 @@
         { KC_NO, C61, C62, C63, C64, C65, C66 }, \
         { C70, C71, C72, C73, C74, C75, C76 }, \
         { C80, C81, C82, C83, C84, C85, C86 }, \
-        { C90, C91, C92, C93, C94, C95, C96 } \
+        { C90, C91, C92, C93, C94, C95, C96 }, \
+        { KC_WH_D, KC_WH_U, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO } \
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/dynamis.h:60]:
```suggestion
        { C90, C91, C92, C93, C94, C95, C96 } \
```

File: keyboards/bbrfkr/dynamis/config.h [hunk 10]
```diff
@@ -26,9 +26,9 @@
 #define PRODUCT dynamis
 
 /* key matrix */
-#define MATRIX_ROWS 10
+#define MATRIX_ROWS 11
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/config.h:29]:
```suggestion
#define MATRIX_ROWS 10
```

File: keyboards/bbrfkr/dynamis/config.h [hunk 11]
```diff
@@ -26,9 +26,9 @@
...
 #define MATRIX_COLS 7
-#define MATRIX_ROW_PINS { B6, B4, D6, D5, D1, C6, B5, D7, D4, D0 }
+#define MATRIX_ROW_PINS { B6, B4, D6, D5, D1, C6, B5, D7, D4, D0, D2 }
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/config.h:31]: Is D2 actually being used here? Or just being used as a place holder? 
```suggestion
#define MATRIX_ROW_PINS { B6, B4, D6, D5, D1, C6, B5, D7, D4, D0 }
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/bbrfkr/dynamis/config.h (hunk 11 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/config.h:31]: D2 is for the encoder. I think that the default via keymap for encoder cannot be defined if this doesn't exist.

File: keyboards/bbrfkr/dynamis/dynamis.c (hunk 6 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.c]: Ok. I fixed it.

File: keyboards/bbrfkr/dynamis/dynamis.c (hunk 7 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/dynamis.c:27]: Thank you. I added the default behavior of encoder.

File: keyboards/bbrfkr/dynamis/config.h (hunk 11 above)
Comment (drashna) [reply keyboards/bbrfkr/dynamis/config.h:31]: You can undefine and redefine the values for the via keymap, actually.  

Which is the ideal option here, since you're creating blank spots in the layout that are ONLY used for the encoders.  


Also, you should use `NO_PIN` instead of assigning a pin here.

Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/config.h:31]: I see. I replaced `D2` to `NO_PIN`.

**Comment by bbrfkr** [comment]
I didn't notice that this pr conflicted. The conflict has been resolved.

File: keyboards/bbrfkr/dynamis/keymaps/via/keymap.c [hunk 12]
```diff
@@ -20,40 +20,77 @@ enum layer_names {
...
+        encoder_cw.pressed = true;
+        encoder_cw.time = (timer_read() | 1);
+        action_exec(encoder_cw);
+    } else {
+        encoder_ccw.pressed = true;
+        encoder_ccw.time = (timer_read() | 1);
+        action_exec(encoder_ccw);
+    }
+    return false;
+}
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/keymaps/via/keymap.c:96]:
```suggestion
```

File: keyboards/bbrfkr/dynamis/config.h [hunk 13]
```diff
@@ -19,9 +19,9 @@
 #include "config_common.h"
 
 /* key matrix */
-#define MATRIX_ROWS 10
+#define MATRIX_ROWS 11
 #define MATRIX_COLS 7
-#define MATRIX_ROW_PINS { B6, B4, D6, D5, D1, C6, B5, D7, D4, D0 }
+#define MATRIX_ROW_PINS { B6, B4, D6, D5, D1, C6, B5, D7, D4, D0, NO_PIN }
```
Comment (drashna) [inline keyboards/bbrfkr/dynamis/config.h:24]: To clarify, the `NO_PIN` is there to facilitate encoder in the matrix, correct? 

If so, then it needs to be reverted.  It's not the proper way to handle this, especially as VIA now supports the encoder map feature. 

https://docs.qmk.fm/#/feature_encoders?id=encoder-map

This allows for proper handling of all keycodes, without special cases, and for it be configured dynamically, by VIA

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/bbrfkr/dynamis/config.h (hunk 13 above)
Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/config.h:24]: Thank you for your comment. But, I cannot find the document that the via support encoder map... So, I don't understand to construct a json file for supporting encoder map feature...

https://www.caniusevia.com/docs/configuring_qmk

Comment (drashna) [reply keyboards/bbrfkr/dynamis/config.h:24]: https://github.com/qmk/qmk_firmware/pull/17734

Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/config.h:24]: Thank you for your sharing! I'll try it...

Comment (bbrfkr) [reply keyboards/bbrfkr/dynamis/config.h]: Ok. Now, this code support encoder map.

image


The json for via is the following;

```
{
    "name": "dynamis",
    "vendorId": "0x6262",
    "productId": "0x0001",
    "lighting": "qmk_rgblight",
    "matrix": {
        "rows": 10,
        "cols": 7
    },
    "layouts": {
        "labels": [
            "ISO Enter",
            "Split Space",
            "Split Left Shift",
            "Split Right Shift",
            "Split Backspace"
        ],
        "keymap": [
            [
                {
                    "x": 15.5
                },
                "5,0\n\n\n4,1",
                "9,0\n\n\n4,1"
            ],
            [
                {
                    "y": 0.25,
                    "x": 2.5
                },
                "0,0",
                "0,1",
                "0,2",
                "0,3",
                "0,4",
                "0,5",
                "0,6",
                "5,6",
                "5,5",
                "5,4",
                "5,3",
                "5,2",
                "5,1",
                {
                    "w": 2
                },
                "5,0\n\n\n4,0"
            ],
            [
                {
                    "x": 2.5,
                    "w": 1.5
                },
                "1,0",
                "1,1",
                "1,2",
                "1,3",
                "1,4",
                "1,5",
                "1,6",
                "6,6",
                "6,5",
                "6,4",
                "6,3",
                "6,2",
                "6,1",
                {
                    "w": 1.5
                },
                "6,0\n\n\n0,0",
                {
                    "x": 4,
                    "w": 1.25,
                    "h": 2,
                    "w2": 1.5,
                    "h2": 1,
                    "x2": -0.25
                },
                "7,0\n\n\n0,1"
            ],
            [
                {
                    "x": 2.5,
                    "w": 1.75
                },
                "2,0",
                "2,1",
                "2,2",
                "2,3",
                "2,4",
                "2,5",
                "2,6",
                "7,6",
                "7,5",
                "7,4",
                "7,3",
                "7,2",
                {
                    "w": 2.25
                },
                "7,0\n\n\n0,0",
                {
                    "x": 3
                },
                "7,1\n\n\n0,1"
            ],
            [
                {
                    "w": 1.25
                },
                "3,0\n\n\n2,1",
                "3,1\n\n\n2,1",
                {
                    "x": 0.25,
                    "w": 2.25
                },
                "3,1\n\n\n2,0",
                "3,2",
                "3,3",
                "3,4",
                "3,5",
                "3,6",
                "8,6",
                "8,5",
                "8,4",
                "8,3",
                "8,2",
                {
                    "w": 2.75
                },
                "8,1\n\n\n3,0",
                {
                    "x": 0.25
                },
                "9,1",
                {
                    "a": 7
                },
                "e0",
                {
                    "x": 0.25,
                    "a": 4,
                    "w": 1.75
                },
                "8,1\n\n\n3,1",
                "8,0\n\n\n3,1"
            ],
            [
                {
                    "x": 2.5,
                    "w": 1.25
                },
                "4,0",
                {
                    "w": 1.25
                },
                "4,1",
                {
                    "w": 1.25
                },
                "4,2",
                {
                    "w": 2.25
                },
                "4,3\n\n\n1,0",
                {
                    "w": 1.25
                },
                "4,4\n\n\n1,0",
                {
                    "w": 2.75
                },
                "4,5\n\n\n1,0",
                {
                    "w": 1.25
                },
                "4,6",
                {
                    "w": 1.25
                },
                "9,6",
                {
                    "w": 1.25
                },
                "9,5",
                {
                    "x": 0.5
                },
                "9,4",
                "9,3",
                "9,2"
            ],
            [
                {
                    "y": 0.25,
                    "x": 6.25,
                    "w": 6.25
                },
                "4,3\n\n\n1,1"
            ]
        ]
    }
}
```

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by bbrfkr** [comment]
Thank you for your review!
